ambry-store/src/main/java/com.github.ambry.store/IndexValue.java

```diff
@@ -13,114 +13,326 @@
+  private final static int OPERATION_TIME_SECS_SIZE_IN_BYTES = 4;
+  private final static int SERVICE_ID_SIZE_IN_BYTES = 2;
+  private final static int CONTAINER_ID_SIZE_IN_BYTES = 2;
+
+  final static int INDEX_VALUE_SIZE_IN_BYTES_V0 =
+      BLOB_SIZE_IN_BYTES + OFFSET_SIZE_IN_BYTES + FLAG_SIZE_IN_BYTES + EXPIRES_AT_MS_SIZE_IN_BYTES_V0
+          + ORIGINAL_MESSAGE_OFFSET_SIZE_IN_BYTES;
 
-  static int Index_Value_Size_In_Bytes =
-      Blob_Size_In_Bytes + Offset_Size_In_Bytes + Flag_Size_In_Bytes + Expires_At_Ms_Size_In_Bytes
-          + Original_Message_Offset_Size_In_Bytes;
+  final static int INDEX_VALUE_SIZE_IN_BYTES_V1 =
+      BLOB_SIZE_IN_BYTES + OFFSET_SIZE_IN_BYTES + FLAG_SIZE_IN_BYTES + EXPIRES_AT_SECS_SIZE_IN_BYTES_V1
+          + ORIGINAL_MESSAGE_OFFSET_SIZE_IN_BYTES + OPERATION_TIME_SECS_SIZE_IN_BYTES +
+          SERVICE_ID_SIZE_IN_BYTES + CONTAINER_ID_SIZE_IN_BYTES;
 
-  private final ByteBuffer value;
+  private long size;
   private Offset offset;
+  private byte flags;
+  private final long expiresAtMs;
+  private long originalMessageOffset;
+  private final int operationTimeInSecs;
+  private final short serviceId;
+  private final short containerId;
+  private final short version;
 
-  IndexValue(String logSegmentName, ByteBuffer value) {
-    if (value.capacity() != Index_Value_Size_In_Bytes) {
-      throw new IllegalArgumentException("Invalid buffer size");
+  /**
+   * Constructs the {@link IndexValue} with the passed in {@link ByteBuffer} and the given {@code version}
+   * @param logSegmentName the log segment name to be used to construct the offset
+   * @param value the {@link ByteBuffer} representation of the {@link IndexValue}
+   * @param version the version of the {@link PersistentIndex}
+   */
+  IndexValue(String logSegmentName, ByteBuffer value, short version) {
+    this.version = version;
+    if (version == PersistentIndex.VERSION_0) {
+      if (value.capacity() != INDEX_VALUE_SIZE_IN_BYTES_V0) {
+        throw new IllegalArgumentException("Invalid buffer size for version 0");
+      }
+      size = value.getLong();
+      offset = new Offset(logSegmentName, value.getLong());
+      flags = value.get();
+      expiresAtMs = value.getLong();
+      originalMessageOffset = value.getLong();
+      operationTimeInSecs = (int) Utils.Infinite_Time;
+      serviceId = 0;
+      containerId = 0;
+    } else if (version == PersistentIndex.VERSION_1) {
+      if (value.capacity() != INDEX_VALUE_SIZE_IN_BYTES_V1) {
+        throw new IllegalArgumentException("Invalid buffer size for version 1");
+      }
+      size = value.getLong();
+      offset = new Offset(logSegmentName, value.getLong());
+      flags = value.get();
+      int expiresAt = value.getInt();
+      expiresAtMs = expiresAt != Utils.Infinite_Time ? (expiresAt * Time.MsPerSec) : Utils.Infinite_Time;
+      originalMessageOffset = value.getLong();
+      operationTimeInSecs = value.getInt();
```

Comment:
Actually I want to make the arg of IndexValue to integere for operationTime. I don't want to keep changing the long to int and then reverting it back to log when returning. So, to be transparent to the caller, I feel that IndexValue should take in just int for operationTime since it stores the integer value to disk.